wagtail/contrib/frontend_cache/utils.py

```diff
@@ -146,6 +146,112 @@ def purge_pages_from_cache(pages, backend_settings=None, backends=None):
+    ):
+        logger.info("Unable to find purge backend for %s", site.hostname)
+        return
+
+    for name, backend in purge_hostname_backends.items():
+        logger.info(f"Purging hostname via backend: {name}")
+        backend.purge_hostname(site.hostname)
+
+        # NOTE: We could potentially return `None` after the first purge here,
+        # but there's technically nothing to stop developers registering
+        # multiple backends for the same hostname, and no way for us to
+        # know what is redundant, or which backend is more important
+
+    for name, backend in purge_everything_backends.items():
+        # This will likely purge more than necessary, but will be more efficient
+        # than cycling through every page in the tree to generate a list of URLs
+
+        # NOTE: It is up to developers to disable 'everything' purges for
+        # backends where it is too disruptive to other services
+        logger.info(f"Purging everything via backend: {name}")
+        backend.purge_everything()
+
+        # NOTE: We could potentially return `None` after the first purge here,
+        # but there's technically nothing to stop developers registering
+        # multiple backends for the same hostname, and no way for us to
+        # know what is redundant, or which backend is more important
+
+    if other_backends:
+        # Count descendants
+        descendant_count = site.root_page.get_descendants().count()
+        logger.info(
+            "Purging %d page URLs for %s",
+            descendant_count,
+            site.hostname,
+        )
+        batch = PurgeBatch()
+        purge_fixed_paths_separately = False
+        for i, page in enumerate(
+            site.root_page.get_descendants()
+            .select_related("locale")
+            .specific(defer=True)
+            .iterator(url_batch_chunk_size),
+            start=1,
+        ):
+            batch.add_page(page)
+            if len(batch.urls) == url_batch_chunk_size:
+                # Purge the current batch
+                batch.purge(backends=other_backends.keys())
+                # Report progress
+                logger.info("Progress: %d/%d", i + 1, descendant_count)
+                # Start a fresh batch
+                batch = PurgeBatch()
+                # For reporting to make sense, fixed paths must be purged separately
+                purge_fixed_paths_separately = True
+
+        if batch.urls and purge_fixed_paths_separately:
+            # Purge the final batch of page urls
+            batch.purge(backends=other_backends.keys())
+            # Report progress
+            logger.info("Progress: %d/%d", descendant_count, descendant_count)
+            # Start a new batch for purging fixed paths
+            batch = PurgeBatch()
+
+        # NOTE: Add hardcoded paths/urls to the current batch
+        additional_paths = getattr(
+            settings, "WAGTAILFRONTENDCACHE_FIXED_SITE_PATHS", []
+        )
```

Comment:
Question: What is the purpose of this setting? It feels like something which should be derived from the pages themselves, rather than hard-coded.

Reply:
Think about when you actually purge an entire site... In many cases, wagtail pages are only usually a part of that. It's highly likely for there to be authentication-related and other views on URLs that aren't defined in Wagtail anywhere. This was a way to account for everything within the same call.

Reply:
Maybe some kind of hook would be better?